whyis/database/elasticstore.py

```diff
@@ -0,0 +1,469 @@
+        resources = {}
+        json_ld = []
+        bnodes = {}
+
+
+        def skolemize_bnodes(x):
+            if isinstance(x, rdflib.Graph):
+                x = x.identifier
+            if isinstance(x, rdflib.BNode):
+                if x not in bnodes:
+                    bnodes[x] = '_:'+uuid4().hex
+                return bnodes[x]
+            else:
+                return x
+            
+        for quad in quads:
+            s, p, o, g = [skolemize_bnodes(x) for x in quad]
+            if g not in graphs:
+                graphs[g] = {"@id":str(g),"@graph":[]}
+                json_ld.append(graphs[g])
+            if (g,s) not in resources:
+                resources[(g,s)] = {"@id":str(s), "@object" : []}
+                graphs[g]['@graph'].append(resources[(g,s)])
+            resource = resources[(g,s)]
+            if str(p) not in resource:
+                resource[str(p)] = []
+            if isinstance(o, rdflib.Literal):
+                obj = {"@value" : str(o)}
+                if o.datatype is not None:
+                    obj['@type'] = str(o.datatype)
+                if o.language is not None:
+                    obj['@lang'] = str(o.language)
+            else:
+                obj = {"@id" : str(o)}
+            resource[str(p)].append(obj)
+            resource['@object'].append(obj) # check if it's already there first?
+
+        json_ld = json.dumps({ "graphs": json_ld })
+
+        return json_ld
+        
+    def addN(self, quads, docid=None, refresh=False):
+        """\
+        Adds each item in the list of statements to a specific context. The
+        quoted argument is interpreted by formula-aware stores to indicate this
+        statement is quoted/hypothetical.
+        """
+        assert self.__open, "The Store must be open."
+
+        json_ld = self.quads_to_jsonld(quads)
+
+        if docid is None:
+            docid = uuid4().hex
+        else:
+            docid = quote_plus(docid)
+
+        r = self.session.put(self.url+'/nanopublication/'+docid+'?refresh='+('wait_for' if refresh else 'false'),data=json_ld)
+
+        if r.status_code != 201:
+            print(r.status_code)
+            print(r.content)
+    
+                
+    def remove(self, spo, context, txn=None):
+        subject, predicate, object = spo
+        assert False, "remove() is not implemented."
+
+    def retire_nanopub(self, uri):
+        """\
+        Removes from the store a nanopublication that the given Resource object refers to.
+        """
+
+        query = {
+          "query": {
+            "nested": {
+              "query": {
+                "term": {
+                  "graphs.@graph.@id": str(uri)
+                }
+              },
+              "path": "graphs.@graph",
+              "score_mode": "avg"
+            }
+          }
+        }
+
+        print("query is", query)
+        self.session.post(self.url+"/_refresh") # Ensure store is up-to-date before reading
```

Comment:
Duplicate code, factor out

Reply:
What are you referring to here? The call to `/_refresh`?

Reply:
Yes.